hexkit/eventpub/__init__.py

```diff
@@ -0,0 +1,17 @@
+# Copyright 2021 - 2022 Universität Tübingen, DKFZ and EMBL
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""Protocol and accociated providers for event publishing."""
```

Comment:
I think there is no requirement that the library is structured the same way as the apps that are using the library.

My intension was that it is easy to find the providers for a given protocol, i.e. the actual protocol class is grouped together with potential providers. However, I am open to other suggestions. However, we shouldn't go to nested I would say otherwise the import statements become unreadable. Another thing to keep in mind is that providers using different technologies (e.g. RabbitMQ vs Apache Kafka) should not go into the same module since they are linked to different dependencies.

Reply:
Suggestions:



- protocol
    - eventpub.py 
    - eventsub.py 
- providers
    - akafka.py
        - KafkaEventSubscriber
        - KafkaEventPublisher



        
- eventpub
    - protocol.py
    - providers
        - akafka.py
            - KafkaEventPublisher
- eventsub
    - protocol.py
    - providers
        - akafka.py
            - KafkaEventSubscriber


from hexkit.eventsub.akafka import KafkaEventSubscriber

Reply:
I am more and more convinced that the second option is better. As you said, flat is better than nested, and generally, the structure of the library is better to understand this way. Which providers exist for which protocol is probably either way better specified in the documentation. Would you agree @Cito?

Reply:
(I have applied the new structure in case you want to have a look. But we can of course undo it again.)

Reply:
You mean option 1 actually, right? Yes, it makes the actual core elements of the library more prominent.

Reply:
Yes, sorry exactly.

Reply:
Ok, then let's use that new structure.